casr/src/bin/casr-afl.rs

```diff
@@ -171,60 +94,43 @@ fn main() -> Result<()> {
     // Init log.
     util::initialize_logging(&matches);
 
-    // Get output dir
-    let output_dir = matches.get_one::<PathBuf>("output").unwrap();
-    if !output_dir.exists() {
-        fs::create_dir_all(output_dir).with_context(|| {
-            format!("Couldn't create output directory {}", output_dir.display())
-        })?;
-    } else if output_dir.read_dir()?.next().is_some() {
-        bail!("Output directory is not empty.");
-    }
-
-    // Get optional gdb fuzz target args.
-    let gdb_argv: Vec<String> = if let Some(argvs) = matches.get_many::<String>("ARGS") {
-        argvs.cloned().collect()
-    } else {
-        Vec::new()
-    };
+    let casr_san = util::get_path("casr-san")?;
+    let casr_gdb = util::get_path("casr-gdb")?;
 
     // Get all crashes.
-    let mut crashes: HashMap<String, AflCrashInfo> = HashMap::new();
+    let mut crashes: HashMap<String, CrashInfo> = HashMap::new();
     for node_dir in fs::read_dir(matches.get_one::<PathBuf>("input").unwrap())? {
         let path = node_dir?.path();
         if !path.is_dir() {
             continue;
         }
 
         // Get crashes from one node.
-        let mut crash_info = AflCrashInfo::default();
+        let mut crash_info = casr::triage::CrashInfo {
+            casr_tool: casr_gdb.clone(),
+            ..Default::default()
+        };
         let cmdline_path = path.join("cmdline");
         if let Ok(cmdline) = fs::read_to_string(&cmdline_path) {
             crash_info.target_args = cmdline.split_whitespace().map(|s| s.to_string()).collect();
             crash_info.at_index = crash_info
                 .target_args
                 .iter()
                 .skip(1)
-                .position(|s| s.contains("@@"));
+                .position(|s| s.contains("@@"))
+                .map(|x| x + 1);
 
             if let Some(target) = crash_info.target_args.first() {
-                if let Ok(buffer) = fs::read(Path::new(target)) {
-                    if let Ok(elf) = goblin::elf::Elf::parse(&buffer) {
-                        for sym in elf.syms.iter() {
-                            if let Some(name) = elf.strtab.get_at(sym.st_name) {
-                                if name.contains("__asan") {
-                                    crash_info.is_asan = true;
-                                    break;
-                                }
-                            }
+                match util::symbols_list(Path::new(target)) {
+                    Ok(list) => {
+                        if list.contains("__asan") {
+                            crash_info.casr_tool = casr_san.clone()
                         }
-                    } else {
-                        error!("Fuzz target: {} must be an ELF executable.", target);
+                    }
+                    Err(e) => {
+                        error!("{e}");
```

Comment:
Does error support string formating? Or we need format! here?

Reply:
yes, error supports string formating